Add section about PyPI malware detection (#11)
psf/fundable-packaging-improvements: FUNDABLES.md

infrastructure, potentially platform services (e.g., SMS), and community
outreach.

### Productionize Malware Detection

> PSF did an [RfP](https://github.com/python/request-for/blob/master/2019-Q4-PyPI/RFP.md) for some security work on [Warehouse](https://github.com/pypa/warehouse) in late 2019, but the goals for [Milestone 2](https://github.com/python/request-for/blob/master/2019-Q4-PyPI/RFP.md#milestone-2---systems-for-automated-detection-of-malicious-uploads) were more ambitious than funding allowed for. The malware detection system is currently in limbo: an interesting prototype with limited practical impact because of the astounding number of false-positives. Here is a sampling of some of the open issues:

Reviewer: We assume that a person reading this page doesn't necessarily have a lot of context on current problems and why it's important to solve them. So here's a suggested change:

```suggestion
Since anyone can upload a package to PyPI, malicious users might upload malware, which would then harm users. To mitigate this risk, PSF [previously obtained funding](https://github.com/python/request-for/blob/master/2019-Q4-PyPI/RFP.md) to add some malware detection [Warehouse](https://github.com/pypa/warehouse) in late 2019, but the goals for [the relevant milestone](https://github.com/python/request-for/blob/master/2019-Q4-PyPI/RFP.md#milestone-2---systems-for-automated-detection-of-malicious-uploads) were more ambitious than funding allowed for. The malware detection system is currently in limbo: an interesting prototype with limited practical impact because of the astounding number of false-positives. To protect users from malware, we still need to:
```

> - [Make Malware Verdicts Auditable](https://github.com/pypa/warehouse/issues/7421) - verdicts are removed once the associated project/package is removed, which makes it impossible to evaluate the efficacy of this system

Reviewer: ```suggestion
- [Make Malware Verdicts Auditable](https://github.com/pypa/warehouse/issues/7421) - Right now, verdicts are removed once the associated project/package is removed. We need to change the backend to retain past verdicts so we can evaluate and improve the efficacy of this system.
```

> - [Detect packages being published with typo'ish names](https://github.com/pypa/warehouse/issues/4998) - No progress was made on a typosquatting check.

Reviewer: Thanks for writing this up @xmunoz!

Since most of the list concentrates on what still needs doing, not on past lapses, I suggest some rephrasings, such as:

```suggestion
- [Detect packages being published with typo'ish names](https://github.com/pypa/warehouse/issues/4998) - Add a typosquatting check.
```

- [Implement a more robust malware detector](https://github.com/pypa/warehouse/issues/7748) - One current check relies on simple pattern matching with YARA. A better approach requires parsing the package code into an AST.
- [YARA rules for setup.py not ignoring comments](https://github.com/pypa/warehouse/issues/7475) - Related to the above issue.

> Additionally, we tried to set up a partnership with [VirusTotal](https://www.virustotal.com/) during the check development to scan every uploaded package. Sadly, we were unable to finalize the details in time to begin implementation. Integration with a 3rd-party virus scanner is low-hanging fruit that could move the needle on PyPI package security.

Reviewer: ```suggestion
We also want to set up a partnership with [VirusTotal](https://www.virustotal.com/) or a similar third-party virus checking service during the check development to scan every uploaded package. Integration with a third-party virus scanner is low-hanging fruit that could move the needle on PyPI package security.
```

> Funding would be used for backend development and security engineering. Ideally, AV integrations would be donated by the vendors.

Reviewer: In this project, I assume we would also need some frontend work (for the admin interface), technical writing (for instance, adding an item to the FAQ), or community coordination and publicity work. What do you think?

Author: I think the amount of front-end work would be limited, because all of that stuff was part of the scope of the RfP. The issue is really that the backend system doesn't work well.

As far as community collaboration: I think it will be difficult to get people to write checks in their free time with no way to validate their results. Even the testpypi database is locked down, making community contribution an uphill battle.

Reviewer: Thanks for the clarification regarding the frontend work!

I get what you're saying about the difficulty of getting user testers before the implementation is deployed. However, in my experience, community coordination and publicity work might include stuff like:

* recruiting users to test uploads AFTER the initial deployment, and verify that we haven't accidentally broken something for their workflow
* talking with the PyPI admins and moderators to check whether the system is working for what they need, and gathering feedback for the developers to use
* figuring out a way to share some limited validation results with users who want to write checks
* publicizing this work to security researchers and maintainers of other package managers so we can swap tips and things to watch out for

Does that make sense?

Author: > - recruiting users to test uploads AFTER the initial deployment, and verify that we haven't accidentally broken something for their workflow

This requires a lot of involvement from PyPI administrator (admin must approve, deploy and enable out new checks in the UI, which is at least a few hours of work), which I'm hesitant to rely on.

> - talking with the PyPI admins and moderators to check whether the system is working for what they need, and gathering feedback for the developers to use

Again, I don't think a tight feedback loop with PyPI administrators is the way forward on this work.

> - figuring out a way to share some limited validation results with users who want to write checks

After sitting on this thought for a few weeks, I think the most robust and fluid check-development experience will come from releasing a canonical dataset for contributors to validate their checks against. Anything short of this will require, IMO, an unrealistic time investment from PyPI administrators.

> - publicizing this work to security researchers and maintainers of other package managers so we can swap tips and things to watch out for
I think this can already happen, but we don't have a formalized process/forum for it.

Reviewer: To respond to some of your specific notes here: We could request funding for PyPI administrators' work, to give PyPI admins time to do the necessary support to make some of these tasks possible.

In general: I've been giving some examples of work that might fall under "technical writing," "project management," "community coordination," "publicity work," etc. that would be useful in order to finalize malware detection and roll it out to production. As you can see from the other items in the fundable projects list, we usually include a few of those things, because they're usually necessary. It sounds like we both agree that this kind of work would be useful, but -- if PSF gets funding to do that work, or during the writing of a grant proposal -- people would need to think a bit further about what those bits of work would specifically be. Therefore I suggest that you change the line to something like
 
```suggestion
Funding would be used for backend development, security engineering, project management, system administration, and publicity to stakeholders. Ideally, AV integrations would be donated by the vendors.
```

If we DON'T add items like this, then it might be an unpleasant surprise to a potential funder when it turns out we need them!
